// devextreme/esm/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js
/**
 * DevExtreme (esm/__internal/grids/grid_core/__tests__/__mock__/model/grid_core.js)
 * Version: 24.2.14
 * Build date: Fri Feb 20 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import $ from "../../../../../../core/renderer";
import {
    DataRowModel
} from "./row/data_row";
const SELECTORS = {
    headerRowClass: "dx-header-row",
    dataRowClass: "dx-data-row",
    groupRowClass: "dx-group-row"
};
export class GridCoreModel {
    constructor(root) {
        this.root = root
    }
    getHeaderCells() {
        return this.root.querySelectorAll(`.${SELECTORS.headerRowClass} > td`)
    }
    getHeaderCell(columnIndex) {
        return this.getHeaderCells()[columnIndex]
    }
    getCellElement(rowIndex, columnIndex) {
        var _this$root$querySelec;
        return null === (_this$root$querySelec = this.root.querySelectorAll(`.${SELECTORS.dataRowClass}`)[rowIndex]) || void 0 === _this$root$querySelec ? void 0 : _this$root$querySelec.querySelectorAll("td")[columnIndex]
    }
    getGroupRows() {
        return this.root.querySelectorAll(`.${SELECTORS.groupRowClass}`)
    }
    apiColumnOption(id, name, value) {
        switch (arguments.length) {
            case 1:
                return this.getInstance().columnOption(id);
            case 2:
                return this.getInstance().columnOption(id, name);
            default:
                this.getInstance().columnOption(id, name, value);
                return
        }
    }
    getHeaderByText(text) {
        return $(Array.from(this.getHeaderCells()).find((el => $(el).text().includes(text))))
    }
    getDataRows() {
        return this.root.querySelectorAll(`.${SELECTORS.dataRowClass}`)
    }
    getDataRow(rowIndex) {
        return new DataRowModel(this.getDataRows()[rowIndex])
    }
}
