// devextreme/cjs/__internal/scheduler/__mock__/appointment_data_accessor.mock.js
/**
 * DevExtreme (cjs/__internal/scheduler/__mock__/appointment_data_accessor.mock.js)
 * Version: 24.2.14
 * Build date: Fri Feb 20 2026
 *
 * Copyright (c) 2012 - 2026 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.mockFieldExpressions = exports.mockAppointmentDataAccessor = void 0;
var _utils = require("../utils");
const mockFieldExpressions = exports.mockFieldExpressions = {
    startDateExpr: "startDate",
    endDateExpr: "endDate",
    startDateTimeZoneExpr: "startDateTimeZone",
    endDateTimeZoneExpr: "endDateTimeZone",
    allDayExpr: "allDay",
    textExpr: "text",
    descriptionExpr: "description",
    recurrenceRuleExpr: "recurrenceRule",
    recurrenceExceptionExpr: "recurrenceException",
    disabledExpr: "disabled",
    visibleExpr: "visible"
};
const mockAppointmentDataAccessor = exports.mockAppointmentDataAccessor = new _utils.AppointmentDataAccessor(mockFieldExpressions, true);
